### PR TITLE
[3.4] refactoring_03: - Dashboard controller

### PR DESCRIPTION
**Pull request by nickAS21**
## Pull Request description

Put your PR description here instead of this sentence.   

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/service/entitiy/dashboard/DefaultTbDashboardService.java [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        }
+    }
+
+    @Override
+    public Dashboard updateDashboardCustomers(TenantId tenantId, Dashboard dashboard, String[] strCustomerIds, SecurityUser user) throws ThingsboardException {
+        ActionType actionType = ActionType.ASSIGNED_TO_CUSTOMER;
+        try {
+
+
+            Set<CustomerId> customerIds = new HashSet<>();
```
Comment (YevhenBondarenko) [inline application/src/main/java/org/thingsboard/server/service/entitiy/dashboard/DefaultTbDashboardService.java:133]: you should do this in controller

File: application/src/main/java/org/thingsboard/server/service/entitiy/SimpleTbEntityService.java [hunk 2]
```diff
@@ -15,11 +15,15 @@
...
+import org.thingsboard.server.common.data.id.EntityId;
 import org.thingsboard.server.service.security.model.SecurityUser;
 
-public interface SimpleTbEntityService<T> {
+public interface SimpleTbEntityService<E extends HasName, I extends EntityId> {
 
-    T save(T entity, SecurityUser user) throws ThingsboardException;
+   E save(E entity, SecurityUser user) throws ThingsboardException;
+
+   void  delete (E entity, I entityId, SecurityUser user) throws ThingsboardException;
```
[the remark this answers is not in the document]
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/SimpleTbEntityService.java:27]: Ok

File: application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java [hunk 3]
```diff
@@ -71,9 +71,14 @@ public <E extends HasName, I extends EntityId> void notifyEntity(TenantId tenant
     public <E extends HasName, I extends EntityId> void notifyDeleteEntity(TenantId tenantId, I entityId, E entity,
                                                                            CustomerId customerId, ActionType actionType,
                                                                            List<EdgeId> relatedEdgeIds,
-                                                                           SecurityUser user, Object... additionalInfo) {
-        logEntityAction(tenantId, entityId, entity, customerId, actionType, user, additionalInfo);
-        sendDeleteNotificationMsg(tenantId, entityId, entity,  relatedEdgeIds);
+                                                                           SecurityUser user,
+                                                                           String body, Object... additionalInfo) {
+        EntityId entityIdForLogEntityAction = entity instanceof Alarm ? ((Alarm)entity).getOriginator() : entityId;
```
[the remark this answers is not in the document]
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java:76]: Ok

File: application/src/main/java/org/thingsboard/server/service/entitiy/SimpleTbEntityService.java [hunk 4]
```diff
@@ -15,11 +15,15 @@
  */
 package org.thingsboard.server.service.entitiy;
 
+import org.thingsboard.server.common.data.HasName;
 import org.thingsboard.server.common.data.exception.ThingsboardException;
+import org.thingsboard.server.common.data.id.EntityId;
 import org.thingsboard.server.service.security.model.SecurityUser;
 
-public interface SimpleTbEntityService<T> {
+public interface SimpleTbEntityService<E extends HasName, I extends EntityId> {
```
[the remark this answers is not in the document]
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/SimpleTbEntityService.java:23]: OK

File: application/src/main/java/org/thingsboard/server/service/entitiy/asset/DefaultTbAssetService.java [hunk 5]
```diff
@@ -70,6 +71,10 @@ public ListenableFuture<Void> delete(Asset asset, SecurityUser user) throws Thin
         }
     }
 
+
+    @Override
+    public void delete(Asset entity, AssetId entityId, SecurityUser user) throws ThingsboardException {}
```
[the remark this answers is not in the document]
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/asset/DefaultTbAssetService.java:76]: Save

File: application/src/main/java/org/thingsboard/server/service/entitiy/customer/DefaultTbCustomerService.java [hunk 6]
```diff
@@ -48,13 +49,15 @@ public Customer save(Customer customer, SecurityUser user) throws ThingsboardExc
         }
     }
 
+
     @Override
-    public void delete(Customer customer, SecurityUser user) throws ThingsboardException {
-        TenantId tenantId = customer.getTenantId();
+    public void delete(Customer customer, CustomerId customerId, SecurityUser user) throws ThingsboardException {
+        TenantId tenantId = user.getTenantId();
```
[the remark this answers is not in the document]
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/customer/DefaultTbCustomerService.java:55]: Ok

File: application/src/main/java/org/thingsboard/server/service/entitiy/dashboard/DefaultTbDashboardService.java [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+                    actionType, user, e, dashboard.getId().toString());
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public Dashboard updateDashboardCustomers(TenantId tenantId, Dashboard dashboard, String[] strCustomerIds, SecurityUser user) throws ThingsboardException {
+        ActionType actionType = ActionType.ASSIGNED_TO_CUSTOMER;
+        try {
+
```
[the remark this answers is not in the document]
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/dashboard/DefaultTbDashboardService.java]: OK

File: application/src/main/java/org/thingsboard/server/service/entitiy/dashboard/DefaultTbDashboardService.java (hunk 1 above)
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/dashboard/DefaultTbDashboardService.java:131]: Disagree, the maximum number of operations should be removed from the controller to the service

File: application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java [hunk 8]
```diff
@@ -68,12 +67,14 @@ public <E extends HasName, I extends EntityId> void notifyEntity(TenantId tenant
     }
 
     @Override
-    public <E extends HasName, I extends EntityId> void notifyDeleteEntity(TenantId tenantId, I entityId, E entity,
+    public <E extends HasName, I extends EntityId> void notifyDeleteEntity(TenantId tenantId, I entityId, E entity, EntityId originatorId,
```
Comment (YevhenBondarenko) [inline application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java:70]: Create own method for alarm (notifyDeleteAlarm)

File: application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java [hunk 9]
```diff
@@ -68,12 +67,14 @@ public <E extends HasName, I extends EntityId> void notifyEntity(TenantId tenant
...
                                                                            CustomerId customerId, ActionType actionType,
                                                                            List<EdgeId> relatedEdgeIds,
-                                                                           SecurityUser user, Object... additionalInfo) {
-        logEntityAction(tenantId, entityId, entity, customerId, actionType, user, additionalInfo);
-        sendDeleteNotificationMsg(tenantId, entityId, entity,  relatedEdgeIds);
+                                                                           SecurityUser user,
```
Comment (YevhenBondarenko) [inline application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java:73]: move to own method with body

File: application/src/main/java/org/thingsboard/server/service/entitiy/dashboard/DefaultTbDashboardService.java [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+            return savedDashboard;
+        } catch (Exception e) {
+            notificationEntityService.notifyEntity(user.getTenantId(), emptyId(EntityType.DASHBOARD), null, null,
+                    actionType, user, e, dashboard.getId().toString());
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public Dashboard updateDashboardCustomers(Dashboard dashboard, String[] strCustomerIds, SecurityUser user) throws ThingsboardException {
```
Comment (YevhenBondarenko) [inline application/src/main/java/org/thingsboard/server/service/entitiy/dashboard/DefaultTbDashboardService.java:128]: updateDashboardCustomers(Dashboard dashboard, List<CustomerId> customerIds, SecurityUser user)

Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/dashboard/DefaultTbDashboardService.java:128]: Ok

File: application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java (hunk 9 above)
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java:73]: ok

File: application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java (hunk 8 above)
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java:70]: ok

File: application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java [hunk 11]
```diff
@@ -126,7 +134,7 @@ public void notifyDeleteDevice(TenantId tenantId, DeviceId deviceId, CustomerId
         gatewayNotificationsService.onDeviceDeleted(device);
         tbClusterService.onDeviceDeleted(device, null);
 
-        notifyDeleteEntity(tenantId, deviceId, device, customerId, ActionType.DELETED, relatedEdgeIds, user, false, additionalInfo);
+        notifyDeleteEntity(tenantId, deviceId, device, customerId, ActionType.DELETED, relatedEdgeIds, user,null, additionalInfo);
```
Comment (YevhenBondarenko) [inline application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java:137]: remove null

File: application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java [hunk 12]
```diff
@@ -73,7 +72,16 @@ public <E extends HasName, I extends EntityId> void notifyDeleteEntity(TenantId
                                                                            List<EdgeId> relatedEdgeIds,
                                                                            SecurityUser user, Object... additionalInfo) {
         logEntityAction(tenantId, entityId, entity, customerId, actionType, user, additionalInfo);
-        sendDeleteNotificationMsg(tenantId, entityId, entity,  relatedEdgeIds);
+        sendDeleteNotificationMsg(tenantId, entityId, entity, relatedEdgeIds);
+    }
+
+    public <E extends HasName, I extends EntityId> void notifyDeleteEntityAlarm(TenantId tenantId, I entityId, E entity, EntityId originatorId,
```
Comment (YevhenBondarenko) [inline application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java:78]: rename to notifyDeleteAlarm

File: application/src/main/java/org/thingsboard/server/service/entitiy/asset/DefaultTbAssetService.java [hunk 13]
```diff
@@ -60,7 +61,8 @@ public ListenableFuture<Void> delete(Asset asset, SecurityUser user) throws Thin
         try {
             List<EdgeId> relatedEdgeIds = findRelatedEdgeIds(tenantId, assetId);
             assetService.deleteAsset(tenantId, assetId);
-            notificationEntityService.notifyDeleteEntity(tenantId, assetId, asset, asset.getCustomerId(), ActionType.DELETED, relatedEdgeIds, user, false, asset.toString());
+            notificationEntityService.notifyDeleteEntity(tenantId, assetId, asset, asset.getCustomerId(), ActionType.DELETED,
+                    relatedEdgeIds, user, null, asset.toString());
```
Comment (YevhenBondarenko) [inline application/src/main/java/org/thingsboard/server/service/entitiy/asset/DefaultTbAssetService.java:65]: should be assetId.toString()

File: application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java (hunk 11 above)
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java:137]: OK

File: application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java (hunk 12 above)
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/DefaultTbNotificationEntityService.java:78]: Ok

File: application/src/main/java/org/thingsboard/server/service/entitiy/asset/DefaultTbAssetService.java (hunk 13 above)
Comment (nickAS21) [reply application/src/main/java/org/thingsboard/server/service/entitiy/asset/DefaultTbAssetService.java:65]: Ok

**Review by YevhenBondarenko — APPROVED** [review]

**Pull request merged**